### PR TITLE
Docs: Add "Tuning max_shards" to Prometheus remote write doc

### PR DESCRIPTION
**Pull request by pracucci**
#### PR Description

At Grafana Labs, in Mimir team we're preparing to migrate some customers (technically Mimir cells) to the new architecture. As part of this migration, we'll send out some communications and we're double checking doc accordingly.

I was checking the Prometheus remote write config and I noticed that we don't give any suggestion on how to tune `max_shards`. Tuning `max_shards` and give it some root for spikes is important to make sure that Alloy can keep up with the backlog in case of a outage (e.g. a networking issue between Alloy and Mimir).

In this PR I propose to add a "Tuning max_shards" section in the Prometheus remote write doc page.

_Please **don't** merge yet, even if you agree with the wording. I want to get thoughts from different people first._

#### Which issue(s) this PR fixes

N/A

#### Notes to the Reviewer

N/A

#### PR Checklist

- [x] Documentation added

### CODE REVIEW

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 1]
```diff
@@ -472,6 +474,41 @@ If the component falls behind more than one third of the data written since the
...
+
+The [`queue_config`](#queue_config-block) block allows to configure `max_shards`. The `max_shards` is the maximum
+number of concurrent shards sending samples to the Prometheus-compatible remote write endpoint.
+For each shard, a single remote write request can send up to `max_samples_per_send` samples.
+
+The maximum throughput that Grafana Alloy can achieve when remote writing is a function of
+`max_shards * max_samples_per_send * <write request latency>`. For example, running Alloy with the
+default configuration of 50 `max_shards` and 2000 `max_samples_per_send`, and assuming the
+average latency of a remote write request is 500ms, the maximum throughput achievable is
+about `50 * 2000 * (1s / 500ms) = 200K samples / s`.
```
Comment (dimitarvdimitrov) [inline docs/sources/reference/components/prometheus.remote_write.md:487]:
```suggestion
`max_shards * max_samples_per_send * <1 / average write request latency>`. For example, running Alloy with the
default configuration of 50 `max_shards` and 2000 `max_samples_per_send`, and assuming the
average latency of a remote write request is 500ms, the maximum throughput achievable is
about `50 * 2000 * (1s / 500ms) = 200K samples / s`.
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 2]
```diff
@@ -472,6 +474,41 @@ If the component falls behind more than one third of the data written since the
...
+Assuming a steady number of series of time, without a high churning rate, a rule of
+thumb to configure `max_shards` to a value 4 times higher than the actual shards utilization.
+You can run the following PromQL query to compute the suggested `max_shards` for each
+remote write endpoint `url`:
+
+```
+clamp_min(
+    (
+        # Measure the actual number of shards used the 90th percentile of time.
+        min by(cluster, agent_hostname, url) (ceil(quantile_over_time(0.9, prometheus_remote_storage_shards{url=~".*grafana.*"}[24h])))
```
Comment (dimitarvdimitrov) [inline docs/sources/reference/components/prometheus.remote_write.md:502]: why the `url` matcher?

File: docs/sources/reference/components/prometheus.remote_write.md (hunk 1 above)
Comment (dimitarvdimitrov) [reply docs/sources/reference/components/prometheus.remote_write.md:487]: so the goal of this paragraph is to arrive at a reasonable max_shards number, right? Can we give a query which solves for max_shards - something like `max_shards = target_samples_per_sec * avg_latency_in_sec / max_samples_per_send`?

**Review by dimitarvdimitrov — APPROVED** [review]
looks good, maybe we can give a rearranged formula, but that's minor

File: docs/sources/reference/components/prometheus.remote_write.md (hunk 2 above)
Comment (pracucci) [reply docs/sources/reference/components/prometheus.remote_write.md:502]: Oh it's a typo! It's how I tested it internally 🤦

Comment (pracucci) [reply docs/sources/reference/components/prometheus.remote_write.md]: Removed in https://github.com/grafana/alloy/pull/1018/commits/5b8feb7d5de1391606d718295dd61c9c9d4bc16c

**Comment by pracucci** [comment]
Feedback I received from @pstibrany on Slack:

- I think explanation makes sense overall.
- there’s a typo: root -> room
- perhaps we could make connection between number of series and scrape interval and resulting final sample rate that agents are scraping, so that reader can make a connection between ingestion rate and remote-write throughput
- I think the text jumps to promql quickly, which is great for copy-paste, but some explanation of used metric and entire query before would be nice.

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 3]
```diff
@@ -171,7 +171,8 @@ Each queue then manages a number of concurrent _shards_ which is responsible
 for sending a fraction of data to their respective endpoints. The number of
 shards is automatically raised if samples are not being sent to the endpoint
 quickly enough. The range of permitted shards can be configured with the
-`min_shards` and `max_shards` arguments.
+`min_shards` and `max_shards` arguments. See "[Tuning `max_shards`](#tuning-max_shards)"
+for more information about how to configure `max_shards`.
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:175]:
```suggestion
`min_shards` and `max_shards` arguments. Refer to  [Tuning `max_shards`](#tuning-max_shards)
for more information about how to configure `max_shards`.
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 4]
```diff
@@ -472,6 +474,41 @@ If the component falls behind more than one third of the data written since the
...
 before being pushed to the remote_write endpoint.
 
+### Tuning `max_shards`
+
+The [`queue_config`](#queue_config-block) block allows to configure `max_shards`. The `max_shards` is the maximum
+number of concurrent shards sending samples to the Prometheus-compatible remote write endpoint.
+For each shard, a single remote write request can send up to `max_samples_per_send` samples.
+
+The maximum throughput that Grafana Alloy can achieve when remote writing is a function of
+`max_shards * max_samples_per_send * <write request latency>`. For example, running Alloy with the
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:484]:
```suggestion
`max_shards * max_samples_per_send * <write request latency>`. For example, running {{< param "PRODUCT_NAME" >}} with the
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 5]
```diff
@@ -472,6 +474,41 @@ If the component falls behind more than one third of the data written since the
 last truncate interval, it is possible for the truncate loop to checkpoint data
 before being pushed to the remote_write endpoint.
 
+### Tuning `max_shards`
+
+The [`queue_config`](#queue_config-block) block allows to configure `max_shards`. The `max_shards` is the maximum
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:479]:
```suggestion
The [`queue_config`](#queue_config-block) block allows you to configure `max_shards`. The `max_shards` is the maximum
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 6]
```diff
@@ -472,6 +474,41 @@ If the component falls behind more than one third of the data written since the
...
+number of concurrent shards sending samples to the Prometheus-compatible remote write endpoint.
+For each shard, a single remote write request can send up to `max_samples_per_send` samples.
+
+The maximum throughput that Grafana Alloy can achieve when remote writing is a function of
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:483]:
```suggestion
The maximum throughput that {{< param "PRODUCT_NAME" >}} can achieve when remote writing is a function of
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 7]
```diff
@@ -472,6 +474,41 @@ If the component falls behind more than one third of the data written since the
...
+`max_shards * max_samples_per_send * <write request latency>`. For example, running Alloy with the
+default configuration of 50 `max_shards` and 2000 `max_samples_per_send`, and assuming the
+average latency of a remote write request is 500ms, the maximum throughput achievable is
+about `50 * 2000 * (1s / 500ms) = 200K samples / s`.
+
+The default `max_shards` configuration is good for most use cases. However, if you run Alloy
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:489]:
```suggestion
The default `max_shards` configuration is good for most use cases. However, if you run {{< param "PRODUCT_NAME" >}}
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 8]
```diff
@@ -472,6 +474,41 @@ If the component falls behind more than one third of the data written since the
...
+at a large scale and each Alloy instance scrapes more than 1 million series, we recommend
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:490]:
```suggestion
at a large scale and each {{< param "PRODUCT_NAME" >}} instance scrapes more than 1 million series, we recommend
```

File: docs/sources/reference/components/prometheus.remote_write.md (hunk 6 above)
Comment (09jvilla) [inline docs/sources/reference/components/prometheus.remote_write.md:483]:
```suggestion
The maximum throughput that Grafana Alloy can achieve when remote writing is equal to 
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 9]
```diff
@@ -472,6 +474,41 @@ If the component falls behind more than one third of the data written since the
...
+increasing `max_shards`.
+
+Assuming a steady number of series of time, without a high churning rate, a rule of
```
[the remark this answers is not in the document]
Comment (09jvilla) [reply docs/sources/reference/components/prometheus.remote_write.md:493]: What happens if time series is not constant or you have a high churn rate? Its not clear what I would do in that case. Should we just scratch this entirely?

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 10]
```diff
@@ -472,6 +474,41 @@ If the component falls behind more than one third of the data written since the
...
+thumb to configure `max_shards` to a value 4 times higher than the actual shards utilization.
+You can run the following PromQL query to compute the suggested `max_shards` for each
```
Comment (09jvilla) [inline docs/sources/reference/components/prometheus.remote_write.md:495]:
```suggestion
Run the following PromQL query to compute the suggested `max_shards` value for each
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 11]
```diff
@@ -472,6 +474,41 @@ If the component falls behind more than one third of the data written since the
...
+remote write endpoint `url`:
+
+```
+clamp_min(
+    (
+        # Measure the actual number of shards used the 90th percentile of time.
```
Comment (09jvilla) [inline docs/sources/reference/components/prometheus.remote_write.md:501]:
```suggestion
        # Calculate the 90th percentile in-use shards over the most recent 24 hour period
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 12]
```diff
@@ -472,6 +474,41 @@ If the component falls behind more than one third of the data written since the
...
+        min by(cluster, agent_hostname, url) (ceil(quantile_over_time(0.9, prometheus_remote_storage_shards[24h])))
+
+        # Add room for spikes.
+        * 4
+    ),
+    # Recommended to run at least 50 max_shards, as in the default configuration.
```
Comment (09jvilla) [inline docs/sources/reference/components/prometheus.remote_write.md:507]:
```suggestion
    # We recommend setting max_shards to a value of no less than 50, as in the default configuration.
```

File: docs/sources/reference/components/prometheus.remote_write.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (pracucci) [reply docs/sources/reference/components/prometheus.remote_write.md]: I've rephrase it to make it simple. I've changed the query to look up 7d, so that it takes in account recurrining churning too.

**Comment by pracucci** [comment]
Thanks @clayton-cornell. Could you re-review it, please?

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 13]
```diff
@@ -472,6 +474,53 @@ If the component falls behind more than one third of the data written since the
...
 before being pushed to the remote_write endpoint.
 
+### Tuning `max_shards`
+
+The [`queue_config`](#queue_config-block) block allows you to configure `max_shards`. The `max_shards` is the maximum
+number of concurrent shards sending samples to the Prometheus-compatible remote write endpoint.
+For each shard, a single remote write request can send up to `max_samples_per_send` samples.
+
+The maximum throughput that {{< param "PRODUCT_NAME" >}} can achieve when remote writing is equal to
+`max_shards * max_samples_per_send * <1 / average write request latency>`. For example, running {{< param "PRODUCT_NAME" >}} with the
```
Comment (thampiotr) [inline docs/sources/reference/components/prometheus.remote_write.md:484]: Nit: any reason why not just `(max_shards * max_samples_per_send) / <average write request latency>`?

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 14]
```diff
@@ -472,6 +474,53 @@ If the component falls behind more than one third of the data written since the
...
+default configuration of 50 `max_shards` and 2000 `max_samples_per_send`, and assuming the
+average latency of a remote write request is 500ms, the maximum throughput achievable is
+about `50 * 2000 * (1s / 500ms) = 200K samples / s`.
+
+The default `max_shards` configuration is good for most use cases, especially if each {{< param "PRODUCT_NAME" >}}
+instance scrapes up to 1 million active series. However, if you run {{< param "PRODUCT_NAME" >}}
+at a large scale and each instance scrapes more than 1 million series, we recommend
+increasing the value of `max_shards`.
```
Comment (thampiotr) [inline docs/sources/reference/components/prometheus.remote_write.md:492]: Nit: it may be worth mentioning why we want to limit the max_shards in the first place.

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 15]
```diff
@@ -472,6 +474,53 @@ If the component falls behind more than one third of the data written since the
 last truncate interval, it is possible for the truncate loop to checkpoint data
 before being pushed to the remote_write endpoint.
 
+### Tuning `max_shards`
```
Comment (thampiotr) [inline docs/sources/reference/components/prometheus.remote_write.md:477]: We may want to also link to official Prometheus tuning docs: https://prometheus.io/docs/practices/remote_write/ 

But I find your explanation a bit more practical.

**Review by thampiotr — APPROVED** [review]
LGTM, just couple of small comments

File: docs/sources/reference/components/prometheus.remote_write.md (hunk 13 above)
Comment (pracucci) [reply docs/sources/reference/components/prometheus.remote_write.md:484]: I think it depends how you read it. I don't have a big preference. I did initially as you suggested, then received [this comment](https://github.com/grafana/alloy/pull/1018#discussion_r1634987255) from Dimitar.

File: docs/sources/reference/components/prometheus.remote_write.md (hunk 14 above)
Comment (pracucci) [reply docs/sources/reference/components/prometheus.remote_write.md:498]: Good point. Done in https://github.com/grafana/alloy/pull/1018/commits/5d48e471226cc9db1729899043d881c5c0dc444a

File: docs/sources/reference/components/prometheus.remote_write.md (hunk 15 above)
Comment (pracucci) [reply docs/sources/reference/components/prometheus.remote_write.md:477]: I read it, but I'm not sure it's worth linking it, at least with regards to `max_shards`. I don't think it's saying much about it.

Comment (thampiotr) [reply docs/sources/reference/components/prometheus.remote_write.md:477]: Not feeling too strong about it, so let's not block the PR on this. 

In general this page could have some 'further reading' section though with links, perhaps people may find their answers in the upstream that may be more up-to-date in practice. @clayton-cornell do we have any thoughts about it?

Comment (clayton-cornell) [reply docs/sources/reference/components/prometheus.remote_write.md:477]: If the external info doesn't really add a lot, then we don't really need to link to it. 

It's easy to add: `Refer to the Prometheus [remote write tuning](https://prometheus.io/docs/practices/remote_write/) documentation for more information.`  at the bottom of this section if it helps the users.

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 16]
```diff
@@ -472,6 +474,59 @@ If the component falls behind more than one third of the data written since the
...
+
+The [`queue_config`](#queue_config-block) block allows you to configure `max_shards`. The `max_shards` is the maximum
+number of concurrent shards sending samples to the Prometheus-compatible remote write endpoint.
+For each shard, a single remote write request can send up to `max_samples_per_send` samples.
+
+{{< param "PRODUCT_NAME" >}} will try to not use too many shards, but if the queue falls behind the remote write
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:483]:
```suggestion
{{< param "PRODUCT_NAME" >}} will try not to use too many shards, but if the queue falls behind, the remote write
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 17]
```diff
@@ -472,6 +474,59 @@ If the component falls behind more than one third of the data written since the
...
+component will increase the number of shards up to `max_shards` to increase throughput. A high number of shards may
+potentially overwhelm the remote endpoint, or increase {{< param "PRODUCT_NAME" >}} memory utilization. For this reason,
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:485]:
```suggestion
potentially overwhelm the remote endpoint or increase {{< param "PRODUCT_NAME" >}} memory utilization. For this reason,
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 18]
```diff
@@ -472,6 +474,59 @@ If the component falls behind more than one third of the data written since the
...
+it's important to tune `max_shards` to a reasonable value which is good enough to keep up with the backlog of data
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:486]:
```suggestion
it's important to tune `max_shards` to a reasonable value that is good enough to keep up with the backlog of data
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 19]
```diff
@@ -472,6 +474,59 @@ If the component falls behind more than one third of the data written since the
...
+default configuration of 50 `max_shards` and 2000 `max_samples_per_send`, and assuming the
+average latency of a remote write request is 500ms, the maximum throughput achievable is
+about `50 * 2000 * (1s / 500ms) = 200K samples / s`.
+
+The default `max_shards` configuration is good for most use cases, especially if each {{< param "PRODUCT_NAME" >}}
+instance scrapes up to 1 million active series. However, if you run {{< param "PRODUCT_NAME" >}}
+at a large scale and each instance scrapes more than 1 million series, we recommend
+increasing the value of `max_shards`.
+
+{{< param "PRODUCT_NAME" >}} exposes few metrics that you can use to monitor the remote write shards:
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:500]:
```suggestion
{{< param "PRODUCT_NAME" >}} exposes a few metrics that you can use to monitor the remote write shards:
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 20]
```diff
@@ -472,6 +474,59 @@ If the component falls behind more than one third of the data written since the
...
+
+* `prometheus_remote_storage_shards` (gauge): The number of shards used for concurrent delivery of metrics to an endpoint.
+* `prometheus_remote_storage_shards_min` (gauge): The minimum number of shards a queue is allowed to run.
+* `prometheus_remote_storage_shards_max` (gauge): The maximum number of a shards a queue is allowed to run.
+* `prometheus_remote_storage_shards_desired` (gauge): The number of shards a queue wants to run to be able to keep up with the amount of incoming metrics.
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:505]:
```suggestion
* `prometheus_remote_storage_shards` (gauge): The number of shards used for concurrent delivery of metrics to an endpoint.
* `prometheus_remote_storage_shards_min` (gauge): The minimum number of shards a queue is allowed to run.
* `prometheus_remote_storage_shards_max` (gauge): The maximum number of shards a queue is allowed to run.
* `prometheus_remote_storage_shards_desired` (gauge): The number of shards a queue wants to run to keep up with the number of incoming metrics.
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 21]
```diff
@@ -472,6 +474,59 @@ If the component falls behind more than one third of the data written since the
...
+
+If you're already running {{< param "PRODUCT_NAME" >}}, a rule of thumb is to set `max_shards` to
+4x shard utilization. Using the metrics explained above, you can run the following PromQL instant query
+to compute the suggested `max_shards` value for each remote write endpoint `url`:
+
+```
+clamp_min(
+    (
+        # Calculate the 90th percentile desired shards over the last 7 days period.
+        # If you're running {{< param "PRODUCT_NAME" >}} since less than 7 days, then
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:515]:
```suggestion
        # If you're running {{< param "PRODUCT_NAME" >}} for less than seven days, then
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 22]
```diff
@@ -472,6 +474,59 @@ If the component falls behind more than one third of the data written since the
...
+* `prometheus_remote_storage_shards_desired` (gauge): The number of shards a queue wants to run to be able to keep up with the amount of incoming metrics.
+
+If you're already running {{< param "PRODUCT_NAME" >}}, a rule of thumb is to set `max_shards` to
+4x shard utilization. Using the metrics explained above, you can run the following PromQL instant query
+to compute the suggested `max_shards` value for each remote write endpoint `url`:
+
+```
+clamp_min(
+    (
+        # Calculate the 90th percentile desired shards over the last 7 days period.
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:514]:
```suggestion
        # Calculate the 90th percentile desired shards over the last seven-day period.
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 23]
```diff
@@ -472,6 +474,59 @@ If the component falls behind more than one third of the data written since the
...
+        # Add room for spikes.
+        * 4
+    ),
+    # We recommend setting max_shards to a value of no less than 50, as in the default configuration.
+    50
+)
+```
+
+If you aren't running {{< param "PRODUCT_NAME" >}} yet, we recommend running it with the default `max_shards`
+and then use the PromQL instant query mentioned above to compute the recommended `max_shards`.
```
Comment (clayton-cornell) [inline docs/sources/reference/components/prometheus.remote_write.md:528]:
```suggestion
and then using the PromQL instant query mentioned above to compute the recommended `max_shards`.
```

File: docs/sources/reference/components/prometheus.remote_write.md [hunk 24]
```diff
@@ -472,6 +474,53 @@ If the component falls behind more than one third of the data written since the
...
+
+If you're already running {{< param "PRODUCT_NAME" >}}, a rule of thumb is to set `max_shards` to
+4x shard utilization. Using the metrics explained above, you can run the following PromQL instant query
+to compute the suggested `max_shards` value for each remote write endpoint `url`:
+
+```
+clamp_min(
+    (
+        # Calculate the 90th percentile desired shards over the last 7 days period.
+        # If you're running {{< param "PRODUCT_NAME" >}} since less than 7 days, then
```
[the remark this answers is not in the document]
Comment (clayton-cornell) [reply docs/sources/reference/components/prometheus.remote_write.md:515]: Yes, it works when it's inside in code blocks (I did a make_docs build to check :-) ): 
image

**Review by clayton-cornell — APPROVED** [review]

**Comment by pracucci** [comment]
@thampiotr Anything else I can do to move this PR forward here?

**Comment by thampiotr** [comment]
> @thampiotr Anything else I can do to move this PR forward here?

No, it's looking good, thank you for adding this!

**Comment by pracucci** [comment]
> > @thampiotr Anything else I can do to move this PR forward here?
> 
> No, it's looking good, thank you for adding this!

Thanks to you!

**Comment by clayton-cornell** [comment]
Can we backport this info to v1.0, v1.1, and v1.2? It looks to me like it should be valid for all releases of Alloy

**Comment by thampiotr** [comment]
> Can we backport this info to v1.0, v1.1, and v1.2? It looks to me like it should be valid for all releases of Alloy.

Yes, I think it can be backported. Do you have capacity to do that or should I try pick it up a bit later?

**Comment by clayton-cornell** [comment]
I can do it. I just wanted to be sure before I added the labels for backporting
